roll/websockets.py

```diff
@@ -0,0 +1,153 @@
+    websocket.connection_made(transport)
+    websocket.connection_open()
+    return websocket
+
+
+class WebsocketProtocol(ProtocolUpgrade):
+
+    __slots__ = ('websocket')
+
+    def __init__(self, request, subprotocols, **params):
+        self.subprotocol = None
+        if subprotocols:
+            ws_protocol = request.headers.get('SEC-WEBSOCKET-PROTOCOL')
+            if ws_protocol:
+                # select a subprotocol
+                client_subprotocols = tuple(
+                    (p.strip() for p in ws_protocol.split(',')))
+                for p in client_subprotocols:
+                    if p in subprotocols:
+                        self.subprotocol = p
+                        break
+
+    def bootstrap(self, protocol):
+        self.websocket = create_websocket(protocol.writer, self.subprotocol)
+        return self.websocket
+
+    def upgrade_response(self, request):
+        headers = []
+
+        def get_header(k):
+            return request.headers.get(k.upper(), '')
+
+        def set_header(k, v):
+            headers.append((k, v))
+
+        try:
+            key = handshake.check_request(get_header)
+            handshake.build_response(set_header, key)
+        except InvalidHandshake:
+            raise RuntimeError('Invalid websocket request')
+
+        if self.subprotocol:
+            set_header('Sec-Websocket-Protocol', self.subprotocol)
+
+        rv = b'HTTP/1.1 101 Switching Protocols\r\n'
+        for k, v in headers:
+            rv += (
+                k.encode('utf-8') + b': ' + v.encode('utf-8') + b'\r\n')
+        rv += b'\r\n'
+        return rv
+
+    def connection_lost(self, protocol, exc):
+        self.websocket.connection_lost(exc)
+        return False  # not EoP
+
+    def data_received(self, protocol, data):
+        # Received data. We refuse the data if the websocket is
+        # already closed. If the websocket is closing, this data
+        # might be part of the closing handshake (closing frame)
+        if self.websocket.state != 3:  # not closed
+            self.websocket.data_received(data)
+        else:
+            # The websocket is closed and we still get data for it
+            # This is an unexpected problem. Let's do nothing
+            # about it
+            pass
+        return True  # EoP
+
+    def write(self, protocol, *args):
+        # We don't need a response.
+        protocol.writer.close()
+        return True  # EoP
+
+
+def websocket(app, path, subprotocols: list=None, **extras: dict):
+
+    def websocket_wrapper(handler):
+
+        async def websocket_handler(request, _, **params):
+            # Handshake and protocol upgrade
+            # This is the part that makes the presence of the protocol
+            # mandatory on the request. How could we fix it ?
+            websocket_upgrade = WebsocketProtocol(
+                request, subprotocols, **params)
+            request.protocol.upgrade_protocol(websocket_upgrade)
```

Comment:
This points to a new door we will not open now but we are discussing in an another thread and would like to implement some time: split the parsing process in three steps: url, headers, body, and allow the user to take control after each of those steps instead of having one function view called when everything is parsed.
But that's another journey! :p

Reply:
This was changed in the https://github.com/pyrates/roll/pull/51
Actually, the "refined" versions of the work i did here are there